Add automation to remove old periodics and presubmits for kubevirt/kubevirt sig jobs

dhiller opened the pull request:
Adds the automation to remove old periodics and presubmits for the sigs.

More details here: https://github.com/dhiller/project-infra/blob/20dc020aaccb802f9e831fd118af0cf825e9270e/robots/pkg/kubevirt/cmd/remove/jobs.go#L55

Job definition here: https://github.com/dhiller/project-infra/blob/20dc020aaccb802f9e831fd118af0cf825e9270e/github/ci/prow-deploy/files/jobs/kubevirt/project-infra/project-infra-periodics.yaml#L635

Please note that the current implementation will only remove the "most recent old jobs", it won't remove jobs for even earlier k8s versions. That would have to be done manually. It is to be discussed whether that is necessary.

Further changes:
* major refactoring to use cobra command
* code cleanup, removal of duplicate code
* make jobs use the new pr-creator image to remove unnecessary boilerplate

/cc @enp0s3 @fgimenez

dhiller commented:
/rehearse

fgimenez commented on robots/pkg/kubevirt/github/github.go:
```diff
@@ -0,0 +1,48 @@
+	"kubevirt.io/project-infra/robots/pkg/kubevirt/log"
+)
+
+func NewGitHubClient(ctx context.Context) *github.Client {
```
Could we return an `err` here and let the caller handle it?

fgimenez commented on robots/pkg/kubevirt/cmd/remove/jobs_test.go:
```diff
@@ -0,0 +1,555 @@
+	}
+	for _, tt := range tests {
+		t.Run(tt.name, func(t *testing.T) {
+			tempDir, err := os.MkdirTemp("", "removeOldJobsIfNewOnesExist-")
```
Not for this PR, but we might consider migrating to something like afero to abstract fs dependency in unit tests https://github.com/spf13/afero

fgimenez commented on robots/pkg/kubevirt/cmd/remove/jobs_test.go:
```diff
@@ -0,0 +1,555 @@
+				},
+			},
+			wantAllJobsExist: true,
+		},
```
Could we add a case where some of the sigs but not all exists?

fgimenez commented on robots/pkg/kubevirt/cmd/remove/jobs_test.go:
```diff
@@ -0,0 +1,555 @@
+	}
+}
+
+func hashFiles(original string, tempCopy string) ([32]byte, [32]byte) {
```
hmm this function has duplication, why don't we have just `hashFile` that takes a file path and returns its `sha256.Sum256`?

fgimenez commented on robots/pkg/kubevirt/cmd/remove/jobs_test.go:
```diff
@@ -0,0 +1,555 @@
+}
+
+func createPresubmitJobForRelease(semver *querier.SemVer, sigName string, alwaysRun, optional, skipReport bool) config.Presubmit {
+	res := config.Presubmit{
```
Nit, this function could just return the `config.Presubmit`, no need for the `res` var

fgimenez reviewed (commented):
Great work! Could you please add a small reference to the new commands in the README?

dhiller commented on robots/cmd/kubevirt/main.go:
```diff
@@ -0,0 +1,11 @@
+package main
+
+import (
+	"kubevirt.io/project-infra/robots/pkg/kubevirt/cmd"
```
Thanks for the catch!

dhiller commented on robots/pkg/kubevirt/cmd/copy/jobs.go:
```diff
@@ -0,0 +1,284 @@
+		allPresubmitJobs[job.Name] = job
+	}
+
+	for _, sigName := range jobconfig.SigNames {
```
That's true! WIll need to think about a better one.

dhiller commented on robots/pkg/kubevirt/cmd/copy/jobs.go:
```diff
@@ -0,0 +1,284 @@
+		log.Log().Panicln(err)
+	}
+	releases = querier.ValidReleases(releases)
+	if len(releases) < 2 {
```
Yes, no magic numbers, thanks for the catch!

dhiller replied:
Thinking about what you said regarding cmd option: how would that help? I mean the number 2 is clearly defined by the requirement of a source and a target release needing to be present. Or do you mean we should not fetch the source and target release, but instead make them a command option so that we can copy jobs from any release to any release?

dhiller replied:
If the latter is what you mean, we would need to fetch them from somewhere else? Where would that happen? In a parent command?

fgimenez replied:
You are totally right, in this case we need them to be always 2.

dhiller commented on robots/pkg/kubevirt/cmd/copy/jobs.go:
```diff
@@ -0,0 +1,284 @@
+			newJob.UtilityConfig.ExtraRefs = append(newJob.UtilityConfig.ExtraRefs, extraRef)
+		}
+
+		for index, envVar := range newJob.Spec.Containers[0].Env {
```
Agree!

dhiller commented:
@fgimenez I've addressed review comments, I haven't squashed the commits as then it's easier to see what I did. Hope you don't mind :slightly_smiling_face:

dhiller commented:
One more note: I need to work on the last command, there I didn't yet use RunE. I need to refactor the args parsing a bit more, so that it returns an error also.

dhiller commented on robots/pkg/kubevirt/cmd/flags/flags.go:
```diff
@@ -0,0 +1,80 @@
+	cmd.PersistentFlags().StringVar(&Options.GitHubEndPoint, FlagGitHubEndpoint, "https://api.github.com/", "GitHub's API endpoint (may differ for enterprise).")
+}
+
+func ParseFlagsOrExit(cmd *cobra.Command, args []string, cmdOpts CommandOptions) {
```
This should return an error, caller should take care.

dhiller commented on robots/pkg/kubevirt/cmd/require/presubmits.go:
```diff
@@ -0,0 +1,168 @@
+
+which makes the job required to pass for merges to occur with tide.
+`, shortUsage, strings.Join(prowjobconfigs.SigNames, ", ")),
+	Run: run,
```
Should also use RunE.

dhiller commented:
Done, PTAL! Thanks!

fgimenez commented on robots/pkg/kubevirt/cmd/copy/jobs.go:
```diff
@@ -0,0 +1,292 @@
+
+func copyPresubmitJobsForNewProvider(jobConfig *config.JobConfig, targetProviderReleaseSemver *querier.SemVer, sourceProviderReleaseSemver *querier.SemVer) (updated bool) {
+	allPresubmitJobs := map[string]config.Presubmit{}
+	for index := range jobConfig.PresubmitsStatic[prowjobconfigs.OrgAndRepoForJobConfig] {
```
`prowjobconfigs` looks very good :+1:

fgimenez reviewed (approved):
/approve

Awesome, thanks! And sorry for the late response, missed the ping yesterday

dhiller commented:
> /approve
> 
> Awesome, thanks! And sorry for the late response, missed the ping yesterday

It's fine, thanks for the review! :blush:
